lib/normalize_grammar.re

```diff
@@ -1,5 +1,78 @@
 open Common;
 
-let normalize = _ast =>
-  raise (Todo);
-  
\ No newline at end of file
+let normalize_to_simple = body => {
+  switch (body) {
+  | Ast_grammar.TOKEN =>
+    Ast_grammar_normalized.ATOM(
+      Ast_grammar_normalized.TOKEN
+    );
+  | Ast_grammar.SYMBOL(name)=>
+    Ast_grammar_normalized.ATOM(
+      Ast_grammar_normalized.SYMBOL(name)
+    );
+  /* Ignore rest */
+  | _ => {
+    print_string("Inside normalize_to_simple\n");
+    raise(Todo);
+  }
+  }
+}
+
+let normalize_to_atom = body => {
+  switch (body) {
+  | Ast_grammar.TOKEN =>
+    Ast_grammar_normalized.TOKEN
+  | Ast_grammar.SYMBOL(name)=>
+    Ast_grammar_normalized.SYMBOL(name)
+  /* Ignore rest */
+  | _ => {
+    print_string("Inside normalize_to_atom\n");
+    raise(Todo);
+  }
+  }
+}
+
+let normalize_body = (rule_body) => {
+  switch(rule_body) {
+  | Ast_grammar.TOKEN =>
+    Ast_grammar_normalized.SIMPLE(
+      Ast_grammar_normalized.ATOM(
+        Ast_grammar_normalized.TOKEN
+      )
+    );
+  | Ast_grammar.SYMBOL(name) =>
+    Ast_grammar_normalized.SIMPLE(
+      Ast_grammar_normalized.ATOM(
+        Ast_grammar_normalized.SYMBOL(name)
+      )
+    );
+  | Ast_grammar.SEQ(bodies) =>
+    Ast_grammar_normalized.SIMPLE(
+      Ast_grammar_normalized.SEQ(
+        List.map(normalize_to_atom, bodies)
+      )
+    );
+  | Ast_grammar.CHOICE(bodies) =>
+    Ast_grammar_normalized.CHOICE(
+      List.map(normalize_to_simple, bodies)
+    );
+  | _ =>
+    Ast_grammar_normalized.SIMPLE(
+      Ast_grammar_normalized.ATOM(
+        Ast_grammar_normalized.TOKEN
+      )
+    );
+  }
+}
+
+let normalize_rule = ((name, rule_body)) => (name, normalize_body(rule_body));
```

Comment:
normalize_body will not just return a normalized_body, it will return also new rules for the intermediates, so you'll need something like

let (this_body, intermediates) = normalize_body(rule_body);
let extra_rules = <do_something_on_intermediates_like_calling_again_normalize_rule_on_it>;
(this_body, extra_rules)

Reply:
I addressed it for A.SEQ, and I did something very similar for A.CHOICE matching. LMK if it ok